### ISSUE TITLE
Add `is_multiple_of`

### ISSUE THREAD

**Issue by folkertdev** [rust-lang/libs-team#404]
# Proposal

## Problem statement

Provide a convenient and panic-free method to check whether one numeric value is a multiple of another.

## Motivating examples or use cases

Checking for "is x a multiple of y" is usually done with the expression `x % y == 0`. However, there are two subtle cases in which this expression does not work as intended:

- if `y == 0`, this operation will perform a division by zero.
- if `y == -1` and `x == i{8, 16, 32, 64, 128}::MIN`, then the operation will overflow.

These cases are unlikely to occur in practice, because determining whether a value is a multiple of 0 or -1 is kind of silly. Nonetheless, the potential panic because of division by zero or overflow is always there. 

I'd also argue that `x % y == 0` isn't intuitive. It's a pattern we learn as programmers but is actually kind of hard to explain to newcomers.

Though I've wanted this functionality before, it came up just now in a PR for miri. Miri's codebase is (rightfully!) averse to operations that can panic. However I know that because the right-hand side in my case is not 0 or -1, the panic is actually unreachable. Nonetheless, I have to pollute my code with unwraps.

## Solution sketch

A version of this for all the `{i, u}*` types

```rust
fn is_multiple_of(lhs: i64, rhs: i64) -> bool {
    match rhs {
        // prevent division by zero
        0 => lhs == 0,
        // prevent overflow when lhs == i64::MIN
        -1 => true, 
        _ => lhs % rhs == 0,
    }
}
```

There might be some slight variations that produce better assembly. I think the name aligns well with `next_multiple_of`.

In some real code:

```rust
// before
if cpusetsize.checked_rem(chunk_size).unwrap() == 0 { // can this panic? who knows!
    ...
}

// after
if cpusetsize.is_multiple_of(chunk_size) { // no panics here
    ...
}
```

## Alternatives

- Just do nothing. This is just a relatively minor quality of life improvement

## Links and related work

- [Swift's isMultiple function](https://developer.apple.com/documentation/swift/int/ismultiple(of:))

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by folkertdev** [comment]
To me one of the upsides of the function as currently proposed is that it just never panics. While the next multiple of zero does not make sense, 0 is absolutely a multiple of 0.

**Comment by joshtriplett** [comment]
I would love to see this.

@folkertdev Do you actually need this for signed types, or would your use cases be covered if this just supported unsigned types (like `div_ceil` and `next_multiple_of`)?

**Comment by folkertdev** [comment]
@joshtriplett  I believe I've personally only ever wanted/used this function for unsigned types, but also I think the behavior of `is_multiple_of` for signed types only has one implementation that makes sense (as opposed to e.g. `(-4).next_multiple_of(4)` which could be 0 or -8)

So at the moment I see no reason not to just include signed types in one fell swoop, but if that is easier/faster then they could be split up of course.

**Comment by joshtriplett** [comment]
Mostly asking for bikeshed avoidance reasons: If we know that unsigned is the primary use case, when we go to discuss this if we have any disagreements over the signed versions we can focus on consensus on the unsigned versions.

**Comment by folkertdev** [comment]
To me, the postcondition is very clear: the function answers whether for its input lhs and rhs there exists an integer `c` such that `lhs == c * rhs`. The case where `rhs == 0` follows naturally from the math.

I'll agree that it is weird, in that with the standard `lhs % rhs == 0` pattern the case `rhs == 0` is nonsensical, but in my opinion that is because using `%` for performing the `is_multiple_of` check is clunky.

---

So I guess the question is whether there are cases where the current `lhs % rhs == 0` is used, where the panic on `rhs == 0` prevents bugs (ignoring that panics are DOS vectors in the sort of code that I usually write). Below are some examples from the `rust/rust` codebase found with grepping for `% ..... == 0`.

Sometimes, a check for whether the rhs is zero is already required for correct logic. In those cases, the change in panicking behavior has no impact on the result of the program

```rust
    // If src type size is an integer multiple of the destination type size then
    // the caller will have calculated a `dst_cap` that is an integer multiple of
    // `src_cap` without remainder.
    if const {
        let src_sz = mem::size_of::<SRC>();
        let dest_sz = mem::size_of::<DEST>();
        // would become dest_sz != 0 && src_sz.is_multiple_of(dest_sz)
        dest_sz != 0 && src_sz % dest_sz == 0
    } {
        return false;
    }
```

Other times, the panic is actually a bug: here i believe the behavior is correct when using `i.is_multiple_of(group_size)`. The hidden panic might be a bug here (or is just unreachable in practice).

```rust
    for (i, ch) in s.chars().filter(|&ch| ch != '_').rev().enumerate() {
        // would become: i > 0 && i.is_multiple_of(group_size)
        if i > 0 && i % group_size == 0 {
            chars.push('_');
        }
        chars.push(ch);
    }
```

In other cases I think the bug would be obvious even without the panic, or in any case just as obvious as any typo in the constant values.

```rust
                    // would become: self.index.is_multiple_of(CHUNK_BITS)
                    if self.index % CHUNK_BITS == 0 {
                        break;
                    }
```

```rust
    const CHUNK_SIZE: usize = 192;

    // Check the properties of `CHUNK_SIZE` and `UNROLL_INNER` that are required
    // for correctness.
    const _: () = assert!(CHUNK_SIZE < 256);
    // would become: assert!(CHUNK_SIZE.is_multiple_of(UNROLL_INNER));
    const _: () = assert!(CHUNK_SIZE % UNROLL_INNER == 0);
```

---

in summary, I think the change in behavior is unlikely to come up, and even more unlikely to cause issues in practice, and actually quite likely to eliminate hidden bugs in current code.

**Comment by shepmaster** [comment]
I actually tried to use `if foo.is_multiple_of(bar)` just yesterday because this idea wormed into my brain 😉 

As a semi-silly, semi-serious idea, you could have a `fn is_multiple_of_alt<const N: TheIntType>(self) -> bool` and then be in the same problematic bucket as methods like `slice::as_chunks`.

**Comment by bluebear94** [comment]
For more prior art, Raku has the `%%` operator, but it fails when the denominator is 0.

**Comment by joshtriplett** [comment]
Option 1: We approve this for unsigned types only, and dodge the question. Odds are, most people won't actually care about the missing methods.

Option 2: We implement this for signed types as well, with the semantic @folkertdev suggests.

These seem like the two most reasonable options for us to decide between. I'm fine with either of these options, with a *mild* preference for the former.

**Comment by kennytm** [comment]
@joshtriplett how is unsigned type able to dodge the question of `x.is_multiple_of(0)`? 0 is definitely unsigned.

**Comment by majormice** [comment]
I'd like to provide a convincing case for including a signed variant of `is_multiple_of`

1. Generic Programming
Without `is_multiple_of`, you can't write generic code that works with both signed and unsigned integers. Developers either restrict themselves to unsigned types or duplicate implementations or define a custom trait.
2. API Asymmetry
Rust already has `next_multiple_of` for signed integers. It is counterintuitive and inconsistent that the standard library can compute the next multiple but lacks a built-in, safe way to check if the current value is a multiple.
3. The common alternative is subtly wrong
`x.checked_rem(y).is_some_and(|r| r == 0)` returns false for `x == iN::MIN, y == -1`  because `checked_rem` returns None due to quotient overflow, but mathematically iN::MIN is a multiple of -1.

I ask the libs‑api team to reconsider including signed variant of the initial proposal

**Comment by folkertdev** [comment]
You should open an ACP here https://github.com/rust-lang/libs-team/issues making your case, there is an issue template for that.

or, well, I guess we can revive this? but currently it would not get picked up by automatic agenda generation.

**Comment by kennytm** [comment]
> or, well, I guess we can revive this? but currently it would not get picked up by automatic agenda generation.

This issue already represents the *accepted* ACP for `unsigned_is_multiple_of`. It would be messy to revive this for `signed_is_multiple_of`. Better create a new issue.
